### ISSUE TITLE
LDAP authentication for multiple domains

### ISSUE THREAD

**Issue by koo3** [gitbucket/gitbucket#563]
What can be multiple domains set in the LDAP authentication?
（If the domain name is different.）

**Comment by ghmer** [comment]
This is, imho, an infrastructural question. Are those domains handled by one LDAP system or several systems?
a) Since all information is stored in one LDAP system you just have to provide the correct ldap query string
b) You want to have a virtual directory / meta-directory that maps your LDAP systems. Once you implemented your VDS option a applies

**Comment by koo3** [comment]
Those domains are handled by several LDAP systems.
I would like to set up multiple LDAP servers GitBucket.

**Comment by hansgru** [comment]
+1
We have several LDAP systems too, even with different queries (not just connection parameters).

Another alternative might be if GitBucket would have support for OAuth, thus letting the OAuth server to handle the different connection types.

**Comment by ghmer** [comment]
Well, to be honest, in my opinion this "feature" will cause more problems than it will solve. 
Just imagine you authenticate using several LDAP systems, and an account is found twice/several times: Which account should be used for authentication?

I don't want to be rude, but if you have several LDAP systems within your company, and your accounts are scattered among these systems, you have an architectural problem in your system-landscape. 
If you are some kind of provider and want to give access to a single application using several authentication systems, combine them to a single source - This is why solutions like Meta Directories or Virtual Directories exist. 
I think you should embrace those solutions if you are in the struggle of authenticating users against several systems.

regards, ghmer

**Comment by hansgru** [comment]
> Well, to be honest, in my opinion this "feature" will cause more problems rather than solve them.
Just imagine you authenticate using several LDAP systems, and an account is found twice/several times: Which account should be used for authentication?

The first one that succeeds, authenticates. This is a common scenario in most companies.
Of course, OAuth could make things more transparent.

> I don't want to be rude, but if you have several LDAP systems within your company, and your accounts are scattered among these systems, you have an architectural problem in your system-landscape. 

Nope :). That's typical for most companies that consist of "independent" branches.

> I think you should embrace those solutions if you are in the struggle of authenticating users against several systems.

I'm in no position to dictate such things: I can only try to convince them to use GitBucket for their own good :).

**Comment by ghmer** [comment]
> The first one that succeeds, authenticates. This is a common scenario in most companies.
> Of course, OAuth could make things more transparent.

No, it isn't, and if it was, I would be ashamed of my profession. This just generates false login attempts and the security guys get annoyed by that stupid application that is generating those. I'm also in no position to decide whether or not this feature request will be implemented, but implementing it this way is wrong. Period!
Using an admin account with access to password attributes (that accounts, by the way, do exist very rarely in the companies I am working for. A GIT service should not need such a technical account) is also a bad idea. 
> Nope :). That's typical for most companies that consist of "independent" branches.

You still don't get my point. That's **exactly** why you would use a Virtual Directory service!

Again, I have no influence on this project, but I think I made my points.

**Comment by takahashi-akio** [comment]
In our company, members are divided into 2 OU folders on our single LDAP system. Since GitBucket can hold a non-recursive LDAP setting, we can't authenticate all of our members. Although this issue originally aims to support multiple systems, this function is useful for our condition.

**Comment by takahashi-akio** [comment]
P.S. See also Redmine's LDAP configuration. It can hold multiple LDAP settings with priority.

**Comment by calid** [comment]
FYI I need this to use gitbucket within my company as well.  We use Active Directory and have separate domains for each region globally, but right now I can't configure gitbucket to search all of them, and I also can't configure recursive search on the forest.

I think this is a common use case especially at large companies.  It also seems a lot of software supports this, for example Jenkins:

![selection_032](https://cloud.githubusercontent.com/assets/494405/6803095/5710bfb8-d203-11e4-94bf-978a2d939d67.png)
